Enterprise drive and output format improvements

kendmerry opened the pull request:
Add some changes, primarily to the Enterprise drive code, to do a few things:
o Make it easier to operate on drives that haven't had a password set yet.
o Make the initial setup code work without erring out on Seagate and HGST drives.
o Add an output mode without the time/date to be a little more readable.
o Make sedutil exit statuses consistent. In particular, it's important to know when we get SP_BUSY status back, so we can take action (reset the drive) to restore communication with the encryption processor.

I have tested this on FreeBSD, using the changes from @amotin with Seagate and HGST SED drives.

Sponsored by: Spectra Logic

JanLuca committed:
Add man page for sedutil-cli

r0m30 committed:
New VC file to ignore

r0m30 committed:
Add Function to query disk attachment type

r0m30 committed:
VC no longer allows targeting of pre 8.1 versions of Windows.

r0m30 committed:
Merge Linux code changes

r0m30 committed:
Change identify to use ata passthrough so to support UASP attached USB drives.

r0m30 committed:
Multistart logic for user auth

r0m30 committed:
Add emable/disable to enable user

r0m30 committed:
Add ACEs to lexicon

r0m30 committed:
Merge branch 'master' of //r0m30nas01/Public/gitrepos/sedutil

kendmerry committed:
Fix output formatting for sedutil-cli --query.

The query output for enterprise drives was missing a newline at the
end.

sedutil/Common/DtaDevEnterprise.cpp:
	In DtaDevEnterprise::puke(), keep track of whether we ended
	with a newline so we can print out an extra newline at the
	end of the output if needed.

kendmerry committed:
Clean up return values and exit statuses in sedutil.

Previously, the code for Enterprise encryption would often return
a non-specific failure (0xff) code which was then used as an exit
value.

In order to implement any sort of error recovery for specific
failure cases outside of sedutil, we need to have a clear
indication for the calling program what the failure is.

So, this replaces all of the instances where sedutil returns 0xff
in the case of failure for Enterprise drives with the actual
failure status.

This will allow the caller to, for instance, try resetting the LUN
if sedutil exits with status 0x3 (SP_BUSY).

sedutil/Common/DtaDevEnterprise.cpp:
	Follow the general pattern from DtaDevOpal.ccp, and save
	return codes in a lastRC variable so they can be passed
	back to the caller, rather than just returning 0xff on
	failure.

sedutil/Common/DtaDevEnterprise.h:
	Change the prototypes for getMaxRanges() and
	getMaxRangesOpal() to return status and take the max ranges
	variable on the command line.

sedutil/Common/DtaSession.cpp:
	In DtaSession::start(), allocate a local lastRC.  Print out
	the return code in the event that starting the session
	fails.

sedutil/Common/sedutil.cpp:
	If --printDefaultPassword fails, exit with the proper
	return code instead of exiting with 0.  This will tell
	the caller why it failed, and therefore whether there
	is something he can do to recover.

kendmerry committed:
Checkpoint a number of sedutil changes:

 o Add an alternate output format, accessible with -x, that
   eliminates the time and log level at the start of lines.
 o Print errors to stderr, and informational output to stdout,
   instead of printing everything to stderr.
 o Fix more instances where errors were lost, and not propagated
   back into exit codes.
 o Allow the MSID to be used as a password for --listLockingRange[s]

Common/DtaDev.cpp,
Common/DtaDevEnterprise.h,
Common/DtaDevGeneric.cpp,
Common/DtaHashPwd.cpp,
Common/DtaSession.cpp:
	Include DtaOptions.h, needed for output format type.

Common/DtaDev.h:
	Add the output_format to the DtaDev class.

Common/DtaDevEnterprise.cpp:
	Allow a NULL / NUL password in listLockingRanges().  This
	allows listing locking ranges when the device has not yet
	been initialized with a password.

	Clean up locking range output when -x (readable) output is
	requested.

	When listing one locking range, exit with a failure if that
	one did not succeed.  When listing all locking ranges, exit
	with a failure only if none succeeded.  It looks like there
	a number of cases where we'll get authentication failures
	for some of the locking ranges.

Common/DtaOptions.cpp:
	Add a -x option to set the output format to sedutilReadable.

	Command line parsing in here does not use getopt(3), but is
	rolled by hand.  This would be a candidate for a revamp.

	Initialize the new RCLog class logging level.

Common/log.h:
	Add a new output class, RLog that behaves a little
	differently than the Log class.  It pays attention to the
	output format, and only prints the time and log level if
	we're using sedutilNormal format.  It also prints errors
	to stderr, and normal output to stdout, instead of printing
	everything to stderr.

	Use the new RLog class by default.

Common/sedutil.cpp:
	Store the output format when creating a new device.

kendmerry committed:
Add the ability to erase a drive without first setting a password.

When using the default password (MSID), the password should not be
hashed when establishing the session with the drive.  To indicate
that sedutil-cli should use the default password (and not hash it),
pass in a NULL string for the password "".

This allows wiping a drive that has not been password protected.

Common/DtaDevEnterprise.cpp:
	In eraseLockingRange(), if the user passes in a NULL string
	for the password, fetch the MSID and setup the session to
	not hash the password sent.

kendmerry committed:
Change sedutil-cli to allow resetting passwords back to the default.

The --revertTPer and --revertNoErase options don't seem to work on
Enterprise drives, and so far the only way I've found to really
revert a drive back to factory defaults is using the PSID.  And the
PSID isn't available on the HGST drives I have at least.  (The
Seagate drives do have it.)

So, this allows setting the password back to the default.

Common/DtaDevEnterprise.cpp:
	In setPassword(), get the default password if either the
	current or new password is NULL/NUL.  Set the new password
	to the MSID without hashing if the user wants to go back to
	the default.

kendmerry committed:
Don't fail initial setup if we can't set all BandMaster passwords.

The Seagate ST6000NM0114 drives seem to deny access to BandMaster2
and up.  It isn't clear why, but it might be that you have to
configure BandMaster1 first before they'll allow access to
subsequent bands.

So, in initLSPUsers(), only return an error if we can't set
BandMaster0.  The others are more optional.

Common/DtaDevEnterprise.cpp:
	In initLSPUsers(), don't return an error if we fail to set
	the password for BandMasterN where N > 0.

kendmerry committed:
Add Spectra copyright to the enterprise files.

r0m30 committed:
Allow R/O unlock without auth to modify to writelock state

cristim commented on Common/DtaDevEnterprise.cpp:
```diff
@@ -747,15 +854,15 @@ uint8_t DtaDevEnterprise::setLockingRange_SUM(uint8_t lockingrange, uint8_t lock
 	if (!password && !userid) { LOG(E) << "Formal Parameters"; }
 	LOG(D1) << "Exiting DtaDevEnterprise::enableUser()";
-	return 0xff;
+	return DTAERROR_INVALID_PARAMETER;
```
I'm confused, shouldn't this be DTAERROR_UNSUPORTED_LOCKING_RANGE?

I've seen some more such inconsistencies below, I'm afraid not to introduce some subtle bugs.

ekoyle commented:
I think this would be a lot easier to review in a series of small pull requsts.

r0m30 committed:
Initial NVMe work, identify works

r0m30 committed:
Increase max output buffer size to 60k for better PBA load performance.

r0m30 committed:
Update for faster PBA loading

cristim commented:
@kendmerry I'd really like to see this merged but I'm having a hard time reviewing it in the current state, and it's been pending for the review for too long.

I wonder, is there anyone who would be willing and able to get this reviewed as it is within a few more days?

Otherwise please consider splitting it in a series of PRs, one for each of the implemented features, which would be easier to review.

r0m30 committed:
Custom scan for Linux.

r0m30 committed:
Updated NVMe support for 4.4+ linux kernels.

kendmerry commented:
I’m on vacation this week, but I may have some time next week to work on splitting it up a bit.  It originally went in as separate changes, so I could just put in PRs for groups of those.

Ken
— 
Ken Merry
ken@kdm.org




> On Jul 5, 2017, at 2:53 PM, Cristian Măgherușan-Stanciu @magheru_san <notifications@github.com> wrote:
> 
> @kendmerry <https://github.com/kendmerry> I'd really like to see this merged but I'm having a hard time reviewing it in the current state, and it's been pending for the review for too long.
> 
> I wonder, is there anyone who would be willing and able to get this reviewed as it is within a few more days?
> 
> Otherwise please consider splitting it in a series of PRs, one for each of the implemented features, which would be easier to review.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/sedutil/sedutil/pull/4#issuecomment-313237636>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AbZy5KMLWTygrrwyRrekyTWlV_uxvFJhks5sLAXzgaJpZM4NkI4H>.
>

cristim commented:
It can wait a week longer, unless someone else is able to get it reviewed earlier in the current state

r0m30 committed:
Update for OS specific scan function

r0m30 committed:
Update Copyright for 2017

r0m30 committed:
Update images to buildroot 2017.05.1
- Change PBA?? to ??bit for easier tab completion
- Remove TinyCore - Rescue now same Kernel/libs as PBA
- Add Rescue64 image - UEFI 64 bit image
- No debug PBAs as debugging is now built into PBA
- create/use buildroot package for sedutil-cli and linuxpba to remove lib version issues
- increase size of PBA images to handle larger kernel
- increase size of Rescue to include PBA images

r0m30 committed:
Create autotools build for Linux

r0m30 committed:
Update PBA.
- add support for NVMe drives
- remove ncurses dependency
- better messages
- update NB Makefile(s) -Wall

r0m30 committed:
Update PBA/rescue build scripts

r0m30 committed:
Add the contrib agreement

r0m30 committed:
Minor cleanup

r0m30 committed:
Fix segfault

r0m30 committed:
Merge branch 'addManPage' of git://github.com/JanLuca/sedutil

r0m30 committed:
Distribute the man page

r0m30 committed:
Fix typo on readme.md PR #137

r0m30 committed:
Try and make the builds a little more tarball friendly.

r0m30 committed:
Fix license text.

kendmerry committed:
Restore printing almost everything to stderr in the default case.

This unbreaks the PBA.  If the user adds the -x option, normal output
will go to stdout and errors will go to stderr.

Common/log.h:
	Add a private output format variable in the RLog class and use it
	to track our output format.  In the destructor, only print to
	stdout if we aren't in the sedutilNormal output format.

kendmerry committed:
Change the default to non-log-style stdout/stderr output for the CLI.

For the Linux PBA, the default is that everything goes to stderr in
log style (prepended with the date).

If the user would like log style output sent to stderr only with the
CLI, the new -l option will turn that on.

kendmerry committed:
Merge remote-tracking branch 'upstream/master' into to_upstream

r0m30 committed:
Merge branch 'to_upstream' of https://github.com/kendmerry/sedutil

r0m30 committed:
fix Hidden class variable

kendmerry committed:
Add support for an alternate scan output format.

This tells the OS-specific module to print the scan output in the alternate
format if one is supported.

Common/DtaOptions.cpp:
	Add -a to the usage for --scan and check for it.

Common/DtaOptions.h:
	Add a new altscanoutput option.

Common/sedutil.cpp:
	Pass the altscanoutput option to diskScan().

linux/DtaDevOS.cpp,
linux/DtaDevOS.h,
windows/DtaDevOS.cpp,
windows/DtaDevOS.h:
	Add an alt_output argument to diskScan that defaults to off.

r0m30 committed:
Add additional error messages to revert

r0m30 committed:
Minor kernel changes, try to fix weird mouse issues

r0m30 committed:
Reduce size of rescue image

r0m30 committed:
Update automake/buildroot package to 1.15.1

r0m30 committed:
Update testsuite and baseline for new output format

r0m30 committed:
Update baseline and TestSuite (Linux)

kendmerry committed:
Merge remote-tracking branch 'upstream/master' into to_upstream

cristim commented:
Closing PR, this fork is now obsolete since DTA resumed active development in the open.
